grpc/federation/cel/url.go

```diff
@@ -0,0 +1,528 @@
+					}
+
+					result, err := url.JoinPath(base, paths...)
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					return types.String(result)
+				},
+			),
+		),
+		BindFunction(
+			createURLName("pathEscape"),
+			OverloadFunc(createURLID("pathEscape_string_string"), []*cel.Type{cel.StringType}, cel.StringType,
+				func(_ context.Context, args ...ref.Val) ref.Val {
+					return types.String(url.PathEscape(string(args[0].(types.String))))
+				},
+			),
+		),
+		BindFunction(
+			createURLName("pathUnescape"),
+			OverloadFunc(createURLID("pathUnescape_string_string"), []*cel.Type{cel.StringType}, cel.StringType,
+				func(_ context.Context, args ...ref.Val) ref.Val {
+					result, err := url.PathUnescape(string(args[0].(types.String)))
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					return types.String(result)
+				},
+			),
+		),
+		BindFunction(
+			createURLName("queryEscape"),
+			OverloadFunc(createURLID("queryEscape_string_string"), []*cel.Type{cel.StringType}, cel.StringType,
+				func(_ context.Context, args ...ref.Val) ref.Val {
+					return types.String(url.QueryEscape(string(args[0].(types.String))))
+				},
+			),
+		),
+		BindFunction(
+			createURLName("queryUnescape"),
+			OverloadFunc(createURLID("queryUnescape_string_string"), []*cel.Type{cel.StringType}, cel.StringType,
+				func(_ context.Context, args ...ref.Val) ref.Val {
+					result, err := url.QueryUnescape(string(args[0].(types.String)))
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					return types.String(result)
+				},
+			),
+		),
+
+		// URL functions
+		BindFunction(
+			createURLName("parse"),
+			OverloadFunc(createURLID("parse_string_url"), []*cel.Type{cel.StringType}, URLType,
+				func(_ context.Context, args ...ref.Val) ref.Val {
+					v, err := url.Parse(string(args[0].(types.String)))
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					return lib.toURLValue(*v)
+				},
+			),
+		),
+		BindFunction(
+			createURLName("parseRequestURI"),
+			OverloadFunc(createURLID("parseRequestURI_string_url"), []*cel.Type{cel.StringType}, URLType,
+				func(_ context.Context, args ...ref.Val) ref.Val {
+					v, err := url.ParseRequestURI(string(args[0].(types.String)))
+					if err != nil {
+						return types.NewErr(err.Error())
+					}
+					return lib.toURLValue(*v)
+				},
+			),
+		),
+		BindMemberFunction(
+			"scheme",
+			MemberOverloadFunc(createURLID("scheme_url_string"), URLType, []*cel.Type{}, cel.StringType,
+				func(_ context.Context, self ref.Val, args ...ref.Val) ref.Val {
+					v := lib.refToGoURLValue(self)
+					return types.String(v.Scheme)
+				},
+			),
+		),
+		BindMemberFunction(
+			"opaque",
+			MemberOverloadFunc(createURLID("opaque_url_string"), URLType, []*cel.Type{}, cel.StringType,
+				func(_ context.Context, self ref.Val, args ...ref.Val) ref.Val {
+					v := self.Value().(*URL).GoURL()
```

Comment:
Should we use `lib.refToGoURLValue` instead? (the same goes for other code as well)

Reply:
I forgot about that! I have corrected it.